compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1694,6 +1694,23 @@ trait Applications extends Compatibility {
   def typedUnApply(tree: untpd.UnApply, selType: Type)(using Context): UnApply =
     throw new UnsupportedOperationException("cannot type check an UnApply node")
 
+  def typedAppliedConstructorType(tree: untpd.Apply)(using Context) = untpd.methPart(tree) match
+      case Select(New(tpt), _) =>
+        val tree1 = typedExpr(tree)
+        val widenSkolemsMap = new TypeMap:
+          def apply(tp: Type) = mapOver(tp.widenSkolem)
+        val preciseTp = widenSkolemsMap(tree1.tpe)
+        val classTp = typedType(tpt).tpe
+        def classSymbolHasOnlyTrackedParameters =
+          classTp.classSymbol.primaryConstructor.paramSymss.flatten.filter(_.isTerm).forall(_.is(Tracked))
+        if !preciseTp.isError && !classSymbolHasOnlyTrackedParameters then
+          report.warning(OnlyFullyDependentAppliedConstructorType(), tree.srcPos)
+        if !preciseTp.isError && (preciseTp frozen_=:= classTp) then
+          report.warning(PointlessAppliedConstructorType(tpt, tree.args, classTp), tree.srcPos)
+        TypeTree(preciseTp)
+      case _ =>
+        throw TypeError(em"Unexpected applied constructor type: $tree")
```

Comment:
Can this ever happen for a parsed program? If not we should throw an assertion instead of a TypeError. But with the refactoring proposed in Typer, we can avoid the pattern match here and hence the error case would be eliminated.